components/profile-controller/controllers/profile_controller.go

```diff
@@ -365,39 +372,109 @@ func (r *ProfileReconciler) mapEventToRequest(o client.Object) []reconcile.Reque
+			stoppedCount = stoppedCount + 1
+			if stoppedCount == r.fileWatcherCount {
+				return
+			}
+		case <-ctx.Done():
+			r.Log.Error(errors.New("Timeout"), "Unable to shutdown all watchers")
+			return
+		}
+	}
+}
+
+// Setup file watcher instances to watch namespace labels files
+func (r *ProfileReconciler) SetupFileWatchers(events chan event.GenericEvent, ctx context.Context, mgr ctrl.Manager) error {
+	r.fileWatcherStopped = make(chan int)
+	// Watch config files with namespace labels. If the files change, trigger
 	// a reconciliation for all Profiles.
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return errors.Wrap(err, "Failed to start file watcher")
-	}
-	if err := watcher.Add(r.DefaultNamespaceLabelsPath); err != nil {
-		return errors.Wrapf(err, "Failed to watch file %s", r.DefaultNamespaceLabelsPath)
-	}
-	events := make(chan event.GenericEvent)
-	go func(watcher *fsnotify.Watcher, reconcileEvents chan event.GenericEvent) {
-		defer watcher.Close()
-		for {
-			select {
-			case fsEvent := <-watcher.Events:
-				if fsEvent.Op != fsnotify.Remove && fsEvent.Op != fsnotify.Write {
-					break
+	for _, namespaceLabelsPath := range r.DefaultNamespaceLabelsPaths {
+		// Right now we are creating separate watchers per file to avoid having to write more complicated
+		// logic to have a single watcher watch multiple files and account for symlink resolution because
+		// we only watch two files max. If we need to watch more files later
+		// we should create a solution to have one watcher watch multiple files.
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			r.ShutdownFileWatchers()
+			return errors.Wrap(err, "Failed to start file watcher")
+		}
+		if err := watcher.Add(namespaceLabelsPath); err != nil {
+			r.ShutdownFileWatchers()
+			return errors.Wrapf(err, "Failed to watch file %s", namespaceLabelsPath)
+		}
+
+		go func(watcher *fsnotify.Watcher, reconcileEvents chan event.GenericEvent, ctx context.Context, namespaceLabelsPath string) {
+			elected := false
+			var mu sync.Mutex
+			go func() {
+				// Convey if the process is the leader
+				for {
+					select {
+					case <-mgr.Elected():
+						mu.Lock()
+						elected = true
+						mu.Unlock()
+						return
+					case <-ctx.Done():
+						return
+					}
 				}
-				// ConfigMaps work with symlinks. See:
-				// https://martensson.io/go-fsnotify-and-kubernetes-configmaps/
-				if fsEvent.Op == fsnotify.Remove {
-					watcher.Remove(fsEvent.Name)
-					watcher.Add(r.DefaultNamespaceLabelsPath)
+			}()
+			for {
+				select {
+				case fsEvent := <-watcher.Events:
+					mu.Lock()
+					if fsEvent.Op != fsnotify.Remove && fsEvent.Op != fsnotify.Write {
+						mu.Unlock()
+						break
+					}
+					// ConfigMaps work with symlinks. See:
+					// https://martensson.io/go-fsnotify-and-kubernetes-configmaps/
+					if fsEvent.Op == fsnotify.Remove {
+						watcher.Remove(fsEvent.Name)
+						watcher.Add(namespaceLabelsPath)
+					}
+					if elected {
```

Comment:
This fix is a new addition. I am happy to separate it into a separate PR if that seems more fitting